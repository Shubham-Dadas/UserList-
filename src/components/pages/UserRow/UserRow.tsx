import React, { Component } from "react";

interface UserProps {
  user: {
    id: number;
    name: string;
    email: string;
    gender: string;
    status: string;
  };
}

class UserRow extends Component<UserProps> {
  render() {
    const { name, email, gender, status } = this.props.user;

    return (
      <tr>
        <td>{name}</td>
        <td>{email}</td>
        <td>{gender}</td>
        <td>{status}</td>
      </tr>
    );
  }
}

export default UserRow;
