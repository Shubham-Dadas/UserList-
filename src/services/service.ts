import axios from "axios";
import { accessToken } from "../constants";

const API_URL = "https://gorest.co.in/public/v2/users";

export const getUsers = async (pageNo: number, noOfData: number) => {
  try {
    const response = await axios.get(API_URL, {
      headers: { Authorization: `Bearer ${accessToken}` },
      params: { page: pageNo, per_page: noOfData },
    });

    const totalUsers = parseInt(response.headers["x-pagination-total"] || response.headers["x-total-count"] || "2000");

    return { data: response.data, totalUsers };
  } catch (error) {
    console.error("Error fetching users:", error);
    return { data: [], totalUsers: 0 }; 
  }
};
