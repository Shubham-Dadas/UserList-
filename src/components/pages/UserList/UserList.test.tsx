import { mount, shallow } from "enzyme";

import React from "react";
import toJson from "enzyme-to-json";

jest.mock("../../../services/service", () => ({
  getUsers: jest.fn(),
}));

import { getUsers } from "../../../services/service";

import UsersList from "./UserList";
import UserRow from "../UserRow/UserRow";

describe("UserList component", () => {
  const users = [
    {
      id: 7705369,
      name: "Shubham Dadas",
      email: "shubham.dadas@15ce.com",
      gender: "male",
      status: "active",
    },
    {
      id: 7704657,
      name: "Bankim Nambeesan",
      email: "nambeesan_bankim@waelchi.example",
      gender: "male",
      status: "active",
    },
    {
      id: 7704656,
      name: "Chaturaanan Malik",
      email: "chaturaanan_malik@wilkinson.example",
      gender: "female",
      status: "inactive",
    },
  ];

  const promise = Promise.resolve({
    data: users,
    headers: { "x-pagination-total": 10 },
  });

  (getUsers as jest.Mock).mockReturnValue(promise);

  it("function call when component mounts", () => {
    const component = shallow(<UsersList />);

    expect(getUsers).toHaveBeenCalled();
    return promise.then(() => {
      component.update();

      // @ts-ignore
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  it("loading users content test", () => {
    const component = mount(<UsersList />);
    return promise.then(() => {
      // @ts-ignore
      expect(component.find("p").text()).toEqual("Loading users...");
    });
  });

  it("userRow render testcase", () => {
    const component = mount(<UsersList />);
    return promise.then(() => {
      // @ts-ignore
      component.update();

      expect(component.find(UserRow).length).toEqual(users.length);

      expect(component.find(UserRow).at(0).props().user).toEqual(users[0]);
    });
  });

  it("displays error message when fetching users fails", () => {
    (getUsers as jest.Mock).mockReturnValue(
      Promise.reject(new Error("Error in fetching users"))
    );

    const component = mount(<UsersList />);

    return new Promise((resolve) => setTimeout(resolve, 0)).then(() => {
      component.update();
      expect(component.find(".error p").text()).toEqual(
        "Error in loading users"
      );
    });
  });
});
