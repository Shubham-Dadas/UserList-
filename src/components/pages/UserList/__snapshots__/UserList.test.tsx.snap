// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList component function call when component mounts 1`] = `
<div
  className="user-list-container"
>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Email
          </th>
          <th>
            Gender
          </th>
          <th>
            Status
          </th>
        </tr>
      </thead>
      <tbody>
        <UserRow
          key="7705369"
          user={
            Object {
              "email": "shubham.dadas@15ce.com",
              "gender": "male",
              "id": 7705369,
              "name": "Shubham Dadas",
              "status": "active",
            }
          }
        />
        <UserRow
          key="7704657"
          user={
            Object {
              "email": "nambeesan_bankim@waelchi.example",
              "gender": "male",
              "id": 7704657,
              "name": "Bankim Nambeesan",
              "status": "active",
            }
          }
        />
        <UserRow
          key="7704656"
          user={
            Object {
              "email": "chaturaanan_malik@wilkinson.example",
              "gender": "female",
              "id": 7704656,
              "name": "Chaturaanan Malik",
              "status": "inactive",
            }
          }
        />
      </tbody>
    </table>
    <Pagination
      currentPage={0}
      handlePageClick={[Function]}
      pageCount={1}
    />
  </div>
</div>
`;
